internal/kafka/internal/clusters/provider.go

```diff
@@ -28,6 +29,8 @@ type Provider interface {
 	RemoveResources(clusterSpec *types.ClusterSpec, syncSetName string) error
 	// GetClusterDNS Get the dns of the cluster
 	GetClusterDNS(clusterSpec *types.ClusterSpec) (string, error)
+	// GetCluster
+	GetCluster(clusterID string) (*v1.Cluster, error)
```

Comment:
I think we have to return a different cluster type than the one that comes from ocm sdk.

This way other provider implementation are not tied to ocm provider.

Reply:
is there a viable option here?

Reply:
Yes.

I feel like we already have the struct already. That's the `types.ClusterSpec`, we could enrich it with missing fields e.g cloud provider and region.

The external id is already there.

Reply:
+1 to @machi1990 comment. The cluster Provider has to be fully independent from the underlying provider types

Reply:
Even if you make an interface to conform to v1.Cluster v1.Cluster should not be returned.

If you are referring to keeping an interface conforming to v1.Cluster and return that interface, I think that:
1. Would still have details/methods specific to OCM which we don't at the cluster provider level
2. I don't think it would scale over time. v1.Cluster is a big type with several methods.

Creating its own interface might make sense as a possible alternative too, aside from the suggested approach before which is using `types.ClusterSpec`

Reply:
I looked into `types.ClusterSpec` usage and my new changes shouldn't affect anything. I addressed your comments above. let me know what you think

Reply:
Thanks @pawelpaszki, just one nit: I think we could simply return `types.ClusterSpec` other than a pointer to `types.ClusterSpec`.

Reply:
> If you are referring to keeping an interface conforming to v1.Cluster and return that interface, I think that:
> 
> 1. Would still have details/methods specific to OCM which we don't want at the cluster provider level
> 2. I don't think it would scale over time. v1.Cluster is a big type with several methods.

The interface doesn't need to have all the methods of v1.Cluster. In the interface we can put only the methods we need.

Reply:
> The interface doesn't need to have all the methods of v1.Cluster. In the interface we can put only the methods we need

Yes. That's why I said creating its own interface might make sense as a possible alternative too, aside from the suggested approach before which is using types.ClusterSpec. But in that case then it does not conform to v1.Cluster

Reply:
> Even if you make an interface to conform to v1.Cluster v1.Cluster should not be returned.
> 

@miguelsorianod Why? If we return the interface, the caller will deal only with the methods that we expose in the interface. He won't have knowledge of what is the original object (unless he uses the reflection, but I don't think we should care of that).

Reply:
> Yes. That's why I said creating its own interface might make sense as a possible alternative too, aside from the suggested approach before which is using types.ClusterSpec. But in that case then the interface does not conform to v1.Cluster.

Yes, I got what you mean. With `conform to v1.Cluster` I didn't mean that we need to publish all the `v1.Cluster` methods, but I meant that if the method we publish conforms to methods present in v1.Cluster, we can keep on returning it.

Reply:
v1.Cluster is an OCM type. We shouldn't return v1.Cluste as it is an OCM type
If we create our own interface then we don't return v1.Cluster. We would be returning another type

Reply:
> v1.Cluster is an OCM type. We shouldn't return v1.Cluster as it is an OCM type. If we create our own interface then we by definition we don't return v1.Cluster. We would be returning another type

I think we are talking about 2 different things:
in the signature, we will return the interface, but the `return` statement can return any object compatible with the interface declared in the signature.
If v1.Cluster is compatible, we can return that too.

Reply:
> > Yes. That's why I said creating its own interface might make sense as a possible alternative too, aside from the suggested approach before which is using types.ClusterSpec. But in that case then the interface does not conform to v1.Cluster.
> 
> Yes, I got what you mean. With `conform to v1.Cluster` I didn't mean that we need to publish all the `v1.Cluster` methods, but I meant that if the method we publish conforms to methods present in v1.Cluster, we can keep on returning it.

By `conform to v1.Cluster` I understood you meant all the methods of it, as that's the terminology usually used when using interfaces, hence my response around that.
Assuming you mean having an interface with its own methods, as I mentioned that it might make sense inhttps://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1530#discussion_r1093471645.
But, I don't think it should conform to specific methods of v1.Cluster as the methods in v1.Cluster contain OCM-specific concepts. If your interface ends up providing methods that only apply to OCM you are defeating the purpose of the Provider concept which is being cloud provider agnostic

Reply:
> By `conform to v1.Cluster` I understood you meant all the methods of it, as that's the terminology usually used when using interfaces, hence my response around that.

You are correct. But that usually means that `an object conforms to an interface`. This time I was talking about creating an interface with a `subset` of the methods of `v1.Cluster` (that's why I said `interface conforms to the object`). If we do that, we can keep returning `v1.Cluster` since it will, obviously, conform to our interface.

At the same time:

1. we won't have to make almost any change to our code since the methods will be the same
2. even if in the `return` statement we return `v1.Cluster`, our code will be decoupled from OCM because we will deal only with our interface, and our interface will expose only methods we are interested in

I don't see any advantage in returning a different object.

Reply:
See my previous comment around why I think we shouldn't blindly imitate the methods from the v1.Cluster

Reply:
We'll go with `types.ClusterSpec` and update the existing code that makes references of that type as suggested in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1530#discussion_r1093461285

Thanks for the input